### PR TITLE
Wp 774 TimePicker support Ethiopia 

### PR DESCRIPTION
**Pull request by DavidUM**
Issue
-------
Make timePicker support Ethiopia time

Fix
----
Current timePicker does not consider that meridiem has some entries more than two -AM & PM-
Fix related logic

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

### CODE REVIEW

File: source/TimePicker.js [hunk 1]
```diff
@@ -419,9 +442,11 @@
 					fmtParams.locale = this.locale;
 				}
 				var merFormatter = new ilib.DateFmt(fmtParams);
-				var am = ilib.Date.newInstance({hour:10});
-				var pm = ilib.Date.newInstance({hour:14});
-				this.meridiems = [merFormatter.format(am), merFormatter.format(pm)];
+				if (merFormatter.meridiems == 'chinese') {
+					merFormatter.meridiems = 'default';
+				}
+
```
Comment (DavidUM) [inline source/TimePicker.js]: Hi @graynorton, this part is locale-specific code which I referred in Jira comment.  
Chinese and Ethiopia are only two locale which has some meridiem states more than AM&PM.
However, in last year, we decided to not display whole meridiem unless there is strong request.

This time, Ethiopia asked to show every meridiem and Chinese did not.
So I need to use locale-specific code.

File: source/TimePicker.js [hunk 2]
```diff
@@ -470,10 +492,11 @@
 					if (this.meridiemEnable === true) {
 						this.createComponent(
 							{classes: 'moon-date-picker-wrap', components:[
-								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', value: valueHours > 12 ? 1 : 0, meridiems: this.meridiems || ['am','pm'], onChange: 'meridiemPickerChanged'},
+								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', meridiems: this.meridiems || [{name: 'AM', start: '00:00', end: '11:59'}, {name: 'PM', start: '12:00', end: '23:59'}], max: this.meridiems.length - 1 || 1, onChange: 'meridiemPickerChanged'},
```
Comment (aarontam) [inline source/TimePicker.js]: Hi @DavidUM, there is no `meridiems` property on this kind by default, so among other things, this will not work if `ilib` is not present (as the `ilib` initialization sets the `meridiems` property). We either need more guard code, or should set the initial value of `meridiems` to the defaults you added here.

File: source/TimePicker.js [hunk 3]
```diff
@@ -628,11 +670,11 @@
 		*/
 		setChildPickers: function (inOld) {
 			if (this.value) {
-				var hour = this.value.getHours();
-				this.$.hour.setValue(hour);
-				this.$.minute.setValue(this.value.getMinutes());
-				if (this.meridiemEnable === true) {
-					this.$.meridiem.setValue(hour > 11 ? 1 : 0);
+				var values = this.calcPickerValues();
+				this.$.hour.setValue(values.hour);
```
Comment (aarontam) [inline source/TimePicker.js]: Let's use the generic setter for this and the following lines i.e. `this.$.hour.set('value', values.hour);`

File: source/TimePicker.js [hunk 4]
```diff
@@ -585,14 +608,29 @@
...
-				// reset the hour to < 12 and then add 12 if it's pm
-				hour = hour%12 + value*12;
-				this.updateHours(hour);
+			var meridiems = this.meridiems,
+				hour = this.$.hour.getValue(),
+				oldIndex = event.old,
+				newIndex = this.$.meridiem.getValue(),
+				offset, start, end, newHour;
+
+			/*jshint -W030 */
```
Comment (aarontam) [inline source/TimePicker.js]: Why is this needed?

File: source/TimePicker.js [hunk 5]
```diff
@@ -53,14 +53,37 @@
...
+		*
+		* @param  {Number} hour - hour between 0 to 23
+		* @param  {Number} minute - minute between 0 to 59
+		* @return  {Number} index - index of this.meridiems
+		* @private
+		*/
+		getMeridiemIndex: function (hour, minute) {
+			var meridiems = this.meridiems;
+			for (var i = 0; i < meridiems.length; i++) {
+				var startArr = meridiems[i]['start'].substring(0,2) + meridiems[i]['start'].substring(3,5),
```
Comment (aarontam) [inline source/TimePicker.js]: Is there a reason why we're splitting on ":" in some places and using `substring` with specific indices in other locations, like this?

File: source/TimePicker.js [hunk 6]
```diff
@@ -470,10 +492,11 @@
 					if (this.meridiemEnable === true) {
 						this.createComponent(
 							{classes: 'moon-date-picker-wrap', components:[
-								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', value: valueHours > 12 ? 1 : 0, meridiems: this.meridiems || ['am','pm'], onChange: 'meridiemPickerChanged'},
+								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', meridiems: this.meridiems || [{name: 'AM', start: '00:00', end: '11:59'}, {name: 'PM', start: '12:00', end: '23:59'}], max: this.meridiems.length - 1 || 1, onChange: 'meridiemPickerChanged'},
 								{name: 'meridiemLabel', content: this.meridiemText, classes: 'moon-date-picker-label moon-divider-text'}
 							]}
 						);
+						this.$.meridiem.value = this.$.meridiem.getMeridiemIndex(values.hour, values.minute);
```
Comment (aarontam) [inline source/TimePicker.js]: Is it possible to set the value when creating the `MeridiemPicker`, rather than after the fact?

File: source/TimePicker.js [hunk 7]
```diff
@@ -585,14 +608,29 @@
...
+			/*jshint +W030 */
+
+			if (newHour != hour) {
+				if (newHour > end) {
+					newHour = end;
+				}
+				this.updateValue('hour', this.value.getHours() + (newHour - hour));
+			} else {
+				// there is only single case. 
+				// Ethiopia calendar has different meridiem status between 06:00 and 06:01
```
Comment (aarontam) [inline source/TimePicker.js]: I'm noticing the meridiem changing between 05:59 and 06:00, instead of 06:00 and 06:01, as stated.

File: source/TimePicker.js (hunk 5 above)
Comment (DavidUM) [inline source/TimePicker.js]: Oh yes. I should change variable name.

And I realize that old code are remained. 
During some git processing, I made mistake

File: source/TimePicker.js (hunk 7 above)
Comment (aarontam) [inline source/TimePicker.js]: Never mind, was in the wrong meridiem (can't read this language!).

File: source/TimePicker.js (hunk 6 above)
Comment (DavidUM) [inline source/TimePicker.js]: We should use `getMeridiemIndex()` to find meridiem value.
However, this method is API of moon.MeridiemPicker.
So I placed this line after creation.

And to ignore notice, I assigned its value instead of using setter.

File: source/TimePicker.js (hunk 7 above)
Comment (DavidUM) [inline source/TimePicker.js]: Right. Neither to me.
I just compared its shape. haha.

Comment (aarontam) [inline source/TimePicker.js]: Ok here's a question for you :smile: When using one of the other meridiems (as I was), is it correct that the meridiem changes between 05:59 and 06:00?

File: source/TimePicker.js (hunk 6 above)
Comment (aarontam) [inline source/TimePicker.js]: Ah right, I guess I was more wondering about the ergonomics and encapsulation here. It seems like we might be able to pass something like an `hour` and a `minute` parameter to `MeridiemPicker` and have it compute its own value internally, rather than set the value of `MeridiemPicker` from another kind, based on this other kind calling a method of `MeridiemPicker` (similarly, on L675). It seems like `getMeridiemIndex` would be more of an internal method for `MeridiemPicker`.

File: source/TimePicker.js (hunk 7 above)
Comment (DavidUM) [inline source/TimePicker.js]: Yes! I leave meridiem info for Ethiopia for you.

{ name: 'ጠዋት',
    start: '00:00',
    end: '05:59' },
  { name: 'ቀትር',
    start: '06:00',
    end: '06:00' },
  { name: 'ከሰዓት',
    start: '06:01',
    end: '11:59' },
  { name: 'ከምሽቱ',
    start: '12:00',
    end: '17:59' },
  { name: 'ከሌሊቱ',
    start: '18:00',
    end: '23:59' } ]

That's so interesting :)

File: source/TimePicker.js [hunk 8]
```diff
@@ -628,35 +668,64 @@
...
+
+			return values;
+		},
+
+		/**
+		* @private
+		*/
+		valueChanged: function (old) {
+			if (typeof ilib !== 'undefined') {
+				this.localeValue = ilib.Date.newInstance({unixtime: this.value.getTime(), timezone: 'local'});
```
Comment (aarontam) [inline source/TimePicker.js]: I think there is a case where we would change the value to `null` (as in to reset the `TimePicker`). Currently this would throw an error.

**Comment by aarontam** [comment]
I found some odd behavior when changing the meridiem value via the picker. Using the "down" button of the meridiem picker, the meridiem value animates back to the same value, while the hour decrements, which seems strange, if not incorrect. Can you verify?

File: source/TimePicker.js [hunk 9]
```diff
@@ -470,10 +492,11 @@
 					if (this.meridiemEnable === true) {
 						this.createComponent(
 							{classes: 'moon-date-picker-wrap', components:[
-								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', value: valueHours > 12 ? 1 : 0, meridiems: this.meridiems || ['am','pm'], onChange: 'meridiemPickerChanged'},
+								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', meridiems: this.meridiems || [{name: 'AM', start: '00:00', end: '11:59'}, {name: 'PM', start: '12:00', end: '23:59'}], max: this.meridiems? this.meridiems.length - 1 : 1, onChange: 'meridiemPickerChanged'},
 								{name: 'meridiemLabel', content: this.meridiemText, classes: 'moon-date-picker-label moon-divider-text'}
 							]}
 						);
+						this.$.meridiem.value = this.$.meridiem.getMeridiemIndex(values.hour, values.minute);
```
Comment (aarontam) [inline source/TimePicker.js]: Just mirroring my previous comment here as it got list in the previous diff:

Ah right, I guess I was more wondering about the ergonomics and encapsulation here. It seems like we might be able to pass something like an hour and a minute parameter to MeridiemPicker and have it compute its own value internally, rather than set the value of MeridiemPicker from another kind, based on this other kind calling a method of MeridiemPicker (similarly, on L675). It seems like getMeridiemIndex would be more of an internal method for MeridiemPicker. At the very least, it seems like we could encapsulate the setting of the value inside the MeridiemPicker, and expose a public method to call, that takes an hour and minute parameter.

File: source/TimePicker.js (hunk 7 above)
Comment (aarontam) [inline source/TimePicker.js]: Thanks! This is super helpful.

File: source/TimePicker.js [hunk 10]
```diff
@@ -585,14 +608,27 @@
...
+			offset = hour - parseInt(meridiems[oldIndex]['start'], 10);
+			start = parseInt(meridiems[newIndex]['start'], 10);
+			end = parseInt(meridiems[newIndex]['end'], 10);
+			newHour = offset + start;
+
+			if (newHour != hour) {
+				if (newHour > end) {
+					newHour = end;
+				}
+				this.updateValue('hour', this.value.getHours() + (newHour - hour));
```
Comment (aarontam) [inline source/TimePicker.js]: It seems like this triggers a value update for `TimePicker`, which eventually calls `setChildPickers` and sets the meridiem value back to the previous value.

File: source/TimePicker.js [hunk 11]
```diff
@@ -585,14 +608,27 @@
...
+				oldIndex = event.old,
+				newIndex = this.$.meridiem.getValue(),
+				offset, start, end, newHour;
 
-			if (this.value) {
-				// value is 0 for am, 1 for pm
-				// reset the hour to < 12 and then add 12 if it's pm
-				hour = hour%12 + value*12;
-				this.updateHours(hour);
+			offset = hour - parseInt(meridiems[oldIndex]['start'], 10);
```
Comment (aarontam) [inline source/TimePicker.js]: Just wanted to confirm that the desired behavior is to preserve the offset from the start of the previous meridiem, when changing meridiems. Is this what occurs in other interfaces for this locale?

cc @ehoogerbeets

File: source/TimePicker.js (hunk 8 above)
Comment (DavidUM) [inline source/TimePicker.js]: Yes you are right.
localeValue has a meaning when this.value is exist.
We do not concern about localeValue when this.value is null

File: source/TimePicker.js (hunk 11 above)
Comment (DavidUM) [inline source/TimePicker.js]: When we have just two locale like AM/PM, hour and minute are not changed when we change meridiem betw'n AM/PM.
For example, AM 11:45 will become PM 11:45.
I though this behavior is general in other locale too.

Yes, we'd better get confrim from @ehoogerbeets

Comment (ehoogerbeets) [inline source/TimePicker.js]: For the Ethiopian time system, you cannot switch the meridiem arbitrarily without updating the time. The system goes like this: 

````
12:00-5:59 morning
6:00       noon
6:01-11:59 afternoon
12:00-5:59 evening
6:00-11:59 night
````

So if the UI is showing 7:00 in the afternoon, and you change the meridiem to evening, I would expect that the time gets updated to 12:00, which is the earliest time in the evening.

Comment (graynorton) [inline source/TimePicker.js]: @ehoogerbeets, are there any other time systems that behave this way, or is it only Ethiopian?

Is there any kind of generic indicator from iLib that the current locale requires this type of behavior (changing time when the meridiem changes)?

Basically I'm looking for the most general solution we can apply to the Moonstone control.

File: source/TimePicker.js [hunk 12]
```diff
@@ -609,17 +645,18 @@
 		/**
 		* @private
 		*/
-		updateHours: function (hour) {
+		updateValue: function (newHour, newMinute) {
 			var valueTime = this.value.getTime();
 
-			this.value.setHours(hour);
+			if (newHour !== undefined) { this.value.setHours(newHour); }
```
Comment (aarontam) [inline source/TimePicker.js]: I think we can change this to be `if (newHour != null) this.value.setHours(newHour);`. Similarly, we can do this for the next line as well.

File: source/TimePicker.js [hunk 13]
```diff
@@ -628,35 +665,64 @@
...
 			if (this.value) {
-				var hour = this.value.getHours();
-				this.$.hour.setValue(hour);
-				this.$.minute.setValue(this.value.getMinutes());
-				if (this.meridiemEnable === true) {
-					this.$.meridiem.setValue(hour > 11 ? 1 : 0);
+				var values = this.calcPickerValues();
+				this.$.hour.set('value', values.hour);
+				this.$.minute.set('value', values.minute);
+				if (this.meridiemEnable === true) {
```
Comment (aarontam) [inline source/TimePicker.js]: I think this can be a simple truthy check i.e. `if (this.meridiemEnable) {`

Comment (DavidUM) [inline source/TimePicker.js]: Oh, I agree. Thanks.

File: source/TimePicker.js [hunk 14]
```diff
@@ -470,10 +487,11 @@
 					if (this.meridiemEnable === true) {
 						this.createComponent(
 							{classes: 'moon-date-picker-wrap', components:[
-								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', value: valueHours > 12 ? 1 : 0, meridiems: this.meridiems || ['am','pm'], onChange: 'meridiemPickerChanged'},
+								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', meridiems: this.meridiems || [{name: 'AM', start: '00:00', end: '11:59'}, {name: 'PM', start: '12:00', end: '23:59'}], max: this.meridiems? this.meridiems.length - 1 : 1, onChange: 'meridiemPickerChanged'},
 								{name: 'meridiemLabel', content: this.meridiemText, classes: 'moon-date-picker-label moon-divider-text'}
 							]}
 						);
+						this.$.meridiem.value = this.getMeridiemIndex(values.hour, values.minute);
```
Comment (aarontam) [inline source/TimePicker.js]: I was thinking more like an `updateValue` method we expose on `MeridiemPicker`, that would accept an `hours` and a `minutes` parameter, and would encapsulate the logic found in `getMeridiemIndex`. I realize it's difficult to properly separate everything without doing a wholesale change, but I would like to reduce the amount of meridiem knowledge the `TimePicker` has, and keep that within `MeridiemPicker`. What do you think about this?

File: source/TimePicker.js [hunk 15]
```diff
@@ -470,10 +487,11 @@
 					if (this.meridiemEnable === true) {
 						this.createComponent(
 							{classes: 'moon-date-picker-wrap', components:[
-								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', value: valueHours > 12 ? 1 : 0, meridiems: this.meridiems || ['am','pm'], onChange: 'meridiemPickerChanged'},
+								{kind:'moon.MeridiemPicker', name:'meridiem', classes:'moon-date-picker-field', meridiems: this.meridiems || [{name: 'AM', start: '00:00', end: '11:59'}, {name: 'PM', start: '12:00', end: '23:59'}], max: this.meridiems? this.meridiems.length - 1 : 1, onChange: 'meridiemPickerChanged'},
```
Comment (aarontam) [inline source/TimePicker.js]: So if `this.meridiems` is not specified, we set the `max` to 1 by default. Do you know how this interacts with the statement on L615: https://github.com/enyojs/moonstone/pull/2170/files#diff-0b78987cba1669c429231298cc6c83acR615

File: source/TimePicker.js [hunk 16]
```diff
@@ -419,9 +414,31 @@
 					fmtParams.locale = this.locale;
 				}
 				var merFormatter = new ilib.DateFmt(fmtParams);
-				var am = ilib.Date.newInstance({hour:10});
-				var pm = ilib.Date.newInstance({hour:14});
-				this.meridiems = [merFormatter.format(am), merFormatter.format(pm)];
+				
+				this.meridiems = merFormatter.getMeridiemsRange(fmtParams);
```
Comment (aarontam) [inline source/TimePicker.js]: Do we need to set this as a property of `TimePicker`, or we can keep this set within just `MeridiemPicker`, so that we only have a single system of record. We are referencing `this.meridiems` on L606 (https://github.com/enyojs/moonstone/pull/2170/files#diff-0b78987cba1669c429231298cc6c83acR606), but we can simply retrieve the value from the event originator.

**Comment by aarontam** [comment]
Hi @DavidUM, thanks for the updates! I made a few comments and also noticed that repeatedly changing the meridiem (via pressing the down button of the meridiem picker), causes the minute to increment, so I believe there is something off with the offset still. Let me know if you are unable to reproduce - for me it was pressing until I got to the ከሰዓት meridiem.

File: source/TimePicker.js (hunk 15 above)
Comment (DavidUM) [inline source/TimePicker.js]: {{this.meridiems}} has default value.
Unless developer sets {{this.meridiems}} to undefined, it will has value.

File: source/TimePicker.js (hunk 16 above)
Comment (DavidUM) [inline source/TimePicker.js]: Thank you. Let me update it from 2.6.0-dev.
Cause I'd like to make at least changes as I can to prevent regression issue.

File: source/TimePicker.js (hunk 11 above)
Comment (ehoogerbeets) [inline source/TimePicker.js]: Yes, the Chinese meridiems are like the Ethiopian ones, should we choose to support them. There are other ones like that as well, but ilib doesn't support them yet.

Ilib has a call to retrieve all the meridiems and their start/end times: DateFmt.DateFmt.getMeridiemsRange() which returns an array of {name:string, start:string, end:string} objects which give the details of what the meridiems are called, when they start in "hours:minutes" format, and when they end, also in "hours:minutes". If you only key off of that, then you can support any meridiems style that ilib supports.

File: source/TimePicker.js (hunk 16 above)
Comment (aarontam) [inline source/TimePicker.js]: Hi @DavidUM, thanks for the updates! I think we can remove this assignment, right?

File: source/TimePicker.js [hunk 17]
```diff
@@ -53,14 +53,73 @@
...
+
+			var merFormatter = new ilib.DateFmt(fmtParams);	
+			this.meridiems = merFormatter.getMeridiemsRange(fmtParams);
+		},
+
+		/**
+		* @private
+		*/
+		create: function() {
+			this.initILib();
```
Comment (aarontam) [inline source/TimePicker.js]: I think for the sake of consistency in 2.5 (we've made `enyo-ilib` a dependency of Moonstone in 2.6, so this is not applicable for 2.6), we may want to guard this based on the existence of `ilib`, as we are doing here: https://github.com/enyojs/moonstone/blob/2.5-upkeep/source/DateTimePickerBase.js#L186. What do you think?

File: source/TimePicker.js [hunk 18]
```diff
@@ -53,14 +53,73 @@
...
+		/**
+		* Set meridiem value based on given hour and minute
+		*
+		* @param  {Number} hour - hour between 0 to 23
+		* @param  {Number} minute - minute between 0 to 59
+		*
+		* @public
+		*/
+		setValueByTime: function (hour, minute) {
+			this.value = this.getMeridiemIndex(hour, minute);
```
Comment (aarontam) [inline source/TimePicker.js]: Should this call the setter instead, i.e. `this.set('value', this.getMeridiemIndex(hour, minute)`? Also, I'm wondering if we can fold `getMeridiemIndex` into this function as I don't think it needs to be a separate, public function - it is logic internal to `MeridiemPicker` that only it should know about.

File: source/TimePicker.js [hunk 19]
```diff
@@ -628,35 +690,64 @@
...
-				var hour = this.value.getHours();
-				this.$.hour.setValue(hour);
-				this.$.minute.setValue(this.value.getMinutes());
-				if (this.meridiemEnable === true) {
-					this.$.meridiem.setValue(hour > 11 ? 1 : 0);
+				var values = this.calcPickerValues();
+				this.$.hour.set('value', values.hour);
+				this.$.minute.set('value', values.minute);
+				if (this.meridiemEnable) {					
+					this.$.meridiem.set('value', this.$.meridiem.getMeridiemIndex(values.hour, values.minute));
```
Comment (aarontam) [inline source/TimePicker.js]: Should this use the `setValueByTime` method instead?

File: source/TimePicker.js (hunk 17 above)
Comment (DavidUM) [inline source/TimePicker.js]: I agree.
From 2.6.0-dev, we assume that we have ilib basically, but in 2.5-upkeep we'd better check it.
I'll add guard code.

File: source/TimePicker.js (hunk 18 above)
Comment (DavidUM) [inline source/TimePicker.js]: We can combine `getMeridiemIndex` into this method.
But value assignment is designed because I do not want to notify it.

Comment (aarontam) [inline source/TimePicker.js]: Hm, so when does `valueChanged` get called?

Comment (DavidUM) [inline source/TimePicker.js]: Let me take a look again. Thanks @aarontam

File: source/TimePicker.js [hunk 20]
```diff
@@ -53,14 +53,67 @@
...
+				start = parseInt(meridiems[i]['start'].substring(0,2) + meridiems[i]['start'].substring(3,5), 10);
+				end = parseInt(meridiems[i]['end'].substring(0,2) + meridiems[i]['end'].substring(3,5), 10);
+				time = hour * 100 + minute;
+
+				if ( start <= time && time <= end) {
+					if (this.value != null) {
+						this.set('value', i);
+					} else {
+						//at initial time, we don't want to trigger meridiemPickerChanged method
+						this.value = i;
```
Comment (aarontam) [inline source/TimePicker.js]: Instead of doing this, what do you think about using a call to silence/unsilence, similar to what is done in `moon.DatePicker`? https://github.com/enyojs/moonstone/blob/2.5-upkeep/source/DatePicker.js#L186.

Comment (DavidUM) [inline source/TimePicker.js]: I had a same idea with you. We can re-factor those controls.

And I also consider using silence in this part but I think using assignment is more cheaper.
If you mind to use assignment, let me fix it with using silence.
Thank you @aarontam

Comment (aarontam) [inline source/TimePicker.js]: I think `silence` is only marginally more expensive (it also consists of an assignment, and adds an increment operation), but, for example, to handle the case where value might be set to `null` later and we want to run `setValueByTime` to fire an `onChange` event, I think we should model the behavior in `DatePicker` so that the behavior is more robust.

Comment (DavidUM) [inline source/TimePicker.js]: I agree and will commit soon. Thank you Aaron.

File: source/TimePicker.js [hunk 21]
```diff
@@ -585,14 +602,32 @@
...
-			if (this.value) {
-				// value is 0 for am, 1 for pm
-				// reset the hour to < 12 and then add 12 if it's pm
-				hour = hour%12 + value*12;
-				this.updateHours(hour);
+			if (meridiems.length == 2) {
+				this.updateValue(startHour + offset);
+			} else {
+				var oldMinute =  this.$.minute.get('value'),
+					endHour = parseInt(newMeridiem['end'], 10),
```
Comment (aarontam) [inline source/TimePicker.js]: Let's use dot notation (i.e. `newMeridiem.end`) where possible. This applies to the following two lines, as well.

File: source/TimePicker.js [hunk 22]
```diff
@@ -53,14 +53,60 @@
...
+		* @param  {Number} hour - hour between 0 to 23
+		* @param  {Number} minute - minute between 0 to 59
+		*
+		* @public
+		*/
+		setValueByTime: function (hour, minute) {
+			var meridiems = this.meridiems,
+				start, end, time;
+			for (var i = 0; i < meridiems.length; i++) {
+				start = parseInt(meridiems[i]['start'].substring(0,2) + meridiems[i]['start'].substring(3,5), 10);
```
Comment (aarontam) [inline source/TimePicker.js]: Let's use dot notation for object properties here and on the following line.

File: source/TimePicker.js [hunk 23]
```diff
@@ -75,13 +121,13 @@
 		* @private
 		*/
 		setupItem: function (inSender, inEvent) {
-			var index = inEvent.index % this.range;
-			this.$.item.setContent(this.meridiems[index]);
+			var index = inEvent.index % this.range || 0;
+			this.$.item.setContent(this.meridiems[index]['name']);
```
Comment (aarontam) [inline source/TimePicker.js]: Dot notation here as well.

**Comment by aarontam** [comment]
Now that we're setting the `value` to be an `ilib.Date` object, that means `ilib` is a hard dependency of these controls; if we go down this path, we should remove all of the `ilib` existence checks and related conditionals.

File: source/DatePicker.js [hunk 24]
```diff
@@ -262,7 +253,7 @@
 					second: valueSeconds,
 					millisecond: valueMilliseconds
 				});
-				this.setValue(this.localeValue.getJSDate());
+				this.setValue(localeValue);
```
Comment (aarontam) [inline source/DatePicker.js]: Is it still necessary to do this, or can we just update the existing `ilib.Date` object?

File: source/TimePicker.js [hunk 25]
```diff
@@ -585,14 +602,32 @@
 		meridiemPickerChanged: function (sender, event) {
 			if(this.syncingPickers) return true;
 
-			var hour = this.$.hour.get('value'),
-				value = event.value;
+			var meridiems = event.originator.get('meridiems'),
+				oldMeridiem = meridiems[event.old],
+				newMeridiem = meridiems[this.$.meridiem.get('value')],
+				oldHour = this.$.hour.get('value'),
+				startHour = parseInt(newMeridiem['start'], 10),
```
Comment (aarontam) [inline source/TimePicker.js]: Dot notation here as well.

**Comment by ryanjduffy** [comment]
ilib is a hard dependency for moonstone (for now at least)

**Comment by aarontam** [comment]
Yep! This is for 2.5 :smile:

File: source/TimePicker.js [hunk 26]
```diff
@@ -585,14 +602,32 @@
 		meridiemPickerChanged: function (sender, event) {
 			if(this.syncingPickers) return true;
 
-			var hour = this.$.hour.get('value'),
-				value = event.value;
+			var meridiems = event.originator.get('meridiems'),
+				oldMeridiem = meridiems[event.old],
```
Comment (aarontam) [inline source/TimePicker.js]: Rather than updating the properties of the `onChange` event fired by `IntegerPicker`, I wonder if we can just encapsulate this logic inside `MeridiemPicker` whenever its value changes (we would have access to the previous value), as we should know the most recent hour and minute (via `setValueByTime`), and come up with a clean way for `TimePicker` to retrieve the hour and minute in this change handler. It seems like the logic here is very meridiem-specific.

File: source/DatePicker.js (hunk 24 above)
Comment (ehoogerbeets) [inline source/DatePicker.js]: ilib was not a hard dependency for moonstone before. In fact Gray was pretty adamant that it should not be. Is there some reason it has to become a hard dependency now?

File: source/TimePicker.js (hunk 26 above)
Comment (DavidUM) [inline source/TimePicker.js]: I agree with your thought. Let me update it.
